sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/AlterTableRenameSuite.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.command.v1
+
+import org.apache.spark.sql.{AnalysisException, Row}
+import org.apache.spark.sql.execution.command
+
+/**
+ * This base suite contains unified tests for the `ALTER TABLE .. RENAME` command that check V1
+ * table catalogs. The tests that cannot run for all V1 catalogs are located in more
+ * specific test suites:
+ *
+ *   - V1 In-Memory catalog: `org.apache.spark.sql.execution.command.v1.AlterTableRenameSuite`
+ *   - V1 Hive External catalog: `org.apache.spark.sql.hive.execution.command.AlterTableRenameSuite`
+ */
+trait AlterTableRenameSuiteBase extends command.AlterTableRenameSuiteBase {
+  test("omit namespace in the destination table") {
+    withNamespaceAndTable("ns", "dst_tbl") { dst =>
+      val src = dst.replace("dst", "src")
+      sql(s"CREATE TABLE $src (c0 INT) $defaultUsing")
+      sql(s"INSERT INTO $src SELECT 0")
+
+      sql(s"ALTER TABLE $src RENAME TO dst_tbl")
```

Comment:
doesn't it work with v2 catalogs?

Reply:
v2 places `dst_tbl` directly to the catalog:
```
      sql(s"SHOW TABLES IN $catalog").show(false)

+---------+---------+-----------+
|namespace|tableName|isTemporary|
+---------+---------+-----------+
|         |dst_tbl  |false      |
+---------+---------+-----------+
```
but v1 to the same namespace:
```
+---------+---------+-----------+
|namespace|tableName|isTemporary|
+---------+---------+-----------+
|ns       |dst_tbl  |false      |
+---------+---------+-----------+
```

Reply:
I think it's intuitive to only rename the table instead of moving namespaces if we see `... RENAME TO single_part_name`. Spark should fill the "new name" with the namespace of the source table before calling the v2 API `TableCatalog.renameTable`.

Reply:
Or we can simply update the API doc of `TableCatalog.renameTable` and say: if the `newIdent` has no namespace, it means the namespace of `oldIdent`. This is more backward compatible as it's a doc-only change.

cc @rdblue

Reply:
I believe the API should be simple and straightforward, and shouldn't add extra logic. If you need to rename a table in a namespace, please, specify full path.

Also v1 has only one level of namespaces but v2 has many. And what if I want to move a table to another namespace like:
```
ALTER TABLE ns0.ns1.ns2.ns3.ns4.ns5.tbl RENAME TO ns0.ns1.ns123.tbl
```
v1 prohibits such moving as it allows to rename a table inside of the same database only.

Reply:
From end-user's perspective, it's terrible to force them to write `ALTER TABLE ns0.ns1.ns2.ns3.ns4.ns5.tbl RENAME TO ns0.ns1.ns2.ns3.ns4.ns5.tbl2` if they want to do a simple table rename ...

Reply:
From end-user's perspective, intuitive behavior should be like in file systems, I guess. `mv /ns0/ns1/ns2/ns3/ns4/ns5/tbl tbl2` will not place `tbl2` to `/ns0/ns1/ns2/ns3/ns4/ns5`, right?

Reply:
> we can simply update the API doc of TableCatalog.renameTable and say: if the newIdent has no namespace, it means the namespace of oldIdent

-1 for this idea. I think this is a mistake.

This would introduce ambiguity in the API that requires the implementer to know how to do something correctly. In DSv2, we avoid those cases by making Spark a little smarter so that we have uniform behavior across sources. Instead of adding documentation, Spark should always pass the correct identifier for the rename.

Reply:
@MaxGekk but SQL is not file system. Can you find another database that requires users to fully specify the qualifiers in RENAME?

> Spark should always pass the correct identifier for the rename.

@rdblue This is actually my first proposal, but we need to define what's "correct identifier". Do you agree that "RENAME TO single_part_name" should only rename the table, but not moving the table to the root namespace?

Reply:
> Can you find another database that requires users to fully specify the qualifiers in RENAME?

I have found yet. Some DBMSs don't allow to move a table to another schema/database (DB2, Oracle). MySQL allows to move to another schema/database (https://dev.mysql.com/doc/refman/8.0/en/rename-table.html). But I haven't found any DBMS which moves a table specified only via table name to the "root" namespace like v2 impl does for now.

Reply:
> This is actually my first proposal, but we need to define what's "correct identifier". Do you agree that "RENAME TO single_part_name" should only rename the table, but not moving the table to the root namespace?

My expectation would be to rename the table without changing its namespace. My point was that whatever we decide, Spark should pass the correct full identifier to the catalog so that it is not up to the catalog to make that choice.

What is the current behavior? Would changing this cause a breaking behavior change?

Reply:
The current behavior is different between v1 and v2. V1 follows the mainstream databases and `RENAME TO single_part_name` simply renames the table. V2 leaves the decision to the catalog implementation and our testing in-memory catalog moves the table to the root namespace.

Reply:
Here is the fix for v2: https://github.com/apache/spark/pull/31594